File: src/application/usecase/user/get_roles.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::application::error::AppError;
use crate::application::usecase::user::{RoleRepository, UserRepository};
use crate::domain;

pub struct GetUserRolesInput {
    pub id: String,
}

pub struct GetUserRolesOutput {
    pub roles: Vec<domain::Role>,
}

#[async_trait]
pub trait GetUserRolesUseCase: Send + Sync {
    async fn execute(&self, input: GetUserRolesInput) -> Result<GetUserRolesOutput, AppError>;
}

pub struct GetUserRolesUseCaseImpl {
    user_repo: Arc<dyn UserRepository>,
    role_repo: Arc<dyn RoleRepository>,
}

impl GetUserRolesUseCaseImpl {
    pub fn new(user_repo: Arc<dyn UserRepository>, role_repo: Arc<dyn RoleRepository>) -> Self {
        GetUserRolesUseCaseImpl {
            user_repo,
            role_repo,
        }
    }
}

#[async_trait]
impl GetUserRolesUseCase for GetUserRolesUseCaseImpl {
    async fn execute(&self, input: GetUserRolesInput) -> Result<GetUserRolesOutput, AppError> {
        let user = self.user_repo.find_by_id(&input.id).await?;
        let mut roles = Vec::with_capacity(user.role_ids.len());
        for rid in &user.role_ids {
            if let Ok(role) = self.role_repo.find_by_id(rid).await {
                roles.push(role);
            }
        }
        Ok(GetUserRolesOutput { roles })
    }
}

File: internal/application/usecase/user/get_roles.go
package user

import (
	"context"

	"go-clean-template/internal/domain"
)

type GetUserRolesInput struct {
	ID string
}

type GetUserRolesOutput struct {
	Roles []*domain.Role
}

type GetUserRolesUseCase interface {
	Execute(ctx context.Context, input *GetUserRolesInput) (*GetUserRolesOutput, error)
}

type getUserRolesUseCase struct {
	userRepo domain.UserRepository
	roleRepo domain.RoleRepository
}

func NewGetUserRolesUseCase(userRepo domain.UserRepository, roleRepo domain.RoleRepository) GetUserRolesUseCase {
	return &getUserRolesUseCase{userRepo: userRepo, roleRepo: roleRepo}
}

func (uc *getUserRolesUseCase) Execute(ctx context.Context, input *GetUserRolesInput) (*GetUserRolesOutput, error) {
	user, err := uc.userRepo.FindByID(ctx, input.ID)
	if err != nil {
		return nil, err
	}
	roles := make([]*domain.Role, 0, len(user.RoleIDs))
	for _, rid := range user.RoleIDs {
		role, err := uc.roleRepo.FindByID(ctx, rid)
		if err != nil {
			continue
		}
		roles = append(roles, role)
	}
	return &GetUserRolesOutput{Roles: roles}, nil
}
